core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -1217,6 +1260,71 @@ private[spark] class TaskSetManager(
   def executorAdded(): Unit = {
     recomputeLocality()
   }
+
+  /**
+   * A class for checking inefficient tasks to be speculated, the inefficient tasks come from
+   * the tasks which may be speculated by the previous strategy.
+   */
+  private[scheduler] class InefficientTaskCalculator {
+    private var allTotalRecordsRead = 0L
+    private var allTotalExecutorRunTime = 0L
+    @volatile private var successTaskProgressThreshold = 0.0D
+    private val runingTasksProgressRate = new ConcurrentHashMap[Long, Double]()
+
+    private[scheduler] def updateTaskProgressThreshold(result: DirectTaskResult[_]): Unit = {
+      var totalRecordsRead = 0L
+      var totalExecutorRunTime = 0L
+      result.accumUpdates.foreach { a =>
+        if (a.name == Some(shuffleRead.RECORDS_READ) ||
+          a.name == Some(input.RECORDS_READ)) {
+          val acc = a.asInstanceOf[LongAccumulator]
+          totalRecordsRead += acc.value
+        } else if (a.name == Some(InternalAccumulator.EXECUTOR_RUN_TIME)) {
+          val acc = a.asInstanceOf[LongAccumulator]
+          totalExecutorRunTime = acc.value
+        }
+      }
+      allTotalRecordsRead += totalRecordsRead
+      allTotalExecutorRunTime += totalExecutorRunTime
+      if (allTotalRecordsRead > 0 && allTotalExecutorRunTime > 0) {
+        successTaskProgressThreshold = allTotalRecordsRead / (allTotalExecutorRunTime / 1000.0)
+      }
+    }
+
+    private[scheduler] def updateRuningTasksProgressRate(
```

Comment:
`updateRuningTaskProcessRate`?

Reply:
Other than this method, all other methods in this class can be `private[TaskSetManager]` to limit visibility.
@Ngone51's comment about `synchronized` access to the state will become clear in that case (since all use from `TaskSetManager` would be within scheduler protected lock)